transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -162,7 +162,7 @@ NativeDatagramPacketArray cleanDatagramPacketArray() {
 
     @Override
     protected void wakeup(boolean inEventLoop) {
-        if (!inEventLoop && WAKEN_UP_UPDATER.compareAndSet(this, 0, 1)) {
+        if (!inEventLoop && wakenUp == 0 && WAKEN_UP_UPDATER.compareAndSet(this, 0, 1)) {
```

Comment:
This seems racy in its current form. The wakeup method maybe called by non-EventLoop threads and this may lead to missed events IIUC. For example:

```
Terms:
ELT = EventLoop Thread
T1 = Application Thread (different from ELT)
T2 = Application Thread (different from ELT and T1)

Scenario:
ELT: blocked on epoll_wait, wakenUp is 0
T1: executes task on EL, calls wakeup(..), CAS wakenUp from 0 to 1, calls Native.eventFdWrite(..)
T2: executes task on EL, calls wakeup(..), reads wakenUp == 1
ELT: epoll_wait is unblocked, processes T1's task, GAS wakenUp from 1 to 0, blocked on epoll_wait
T2: returns out of wakeup(..)
```

In this scenario it is unknown if T2's task will be processed.

Reply:
@Scottmitch looking just at this change in isolation I'm fairly certain it can't introduce any race which wasn't already there. I think the case you describe is handled by the additional check in the `run()` method following the call to `epollWait(...)`.

Reply:
@njhill can you clarify which check you are talking about ? I think it is actually "covered" by:

https://github.com/netty/netty/blob/0c43d88f965de22ff162663b20168002e7c14ad8/transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java#L240

But from your comment it seems you are talking about something else :)

Reply:
@Scottmitch could you elaborate?  I haven't looked _super_ carefully at the ELT logic, because logically any bug like the one you are suggesting should be present with or without this change.  If you replace your T2 behaviour with _only_ a CAS, so that it performs a CAS(1, 1) and returns false, the logical semantics are the same for this step, I'm pretty sure?  

Guarding a CAS with a volatile read should only logically modify the visibility semantics (which become only LoadLoad rather than StoreLoad+StoreStore)

In this case, looking briefly at the ELT code, we check the contents of the queue _after_ the CAS inside of `epollWait()` (and `select()`), so I think we're good more specifically.  But I may be missing something since I haven't done exhaustive analysis of the ELT behaviour.

Reply:
@normanmaurer yes I think you're right :) Either way it would have to have been a (major) pre-existing bug...

Reply:
@belliottsmith 
> Guarding a CAS with a volatile read should only logically modify the visibility semantics (which become only LoadLoad rather than StoreLoad+StoreStore)

A volatile load is a LoadLoad + StoreLoad

Reply:
it depends upon the queue implementation. I don't think JCTools MPSC queue guarantees that insertion from one thread will be visible to another thread immediately, and IIUC maybe possible for `hasTasks()` to return false in this scenario. CAS/GAS on the WAKEN_UP_UPDATER will provide a full memory barrier between T2 and ELT and may provide a happens before relationships between the threads. Here are the potential scenarios below:

Scenario | CAS on T2 in wakeUp(..) | GAS(0) on ELT in run() | result
------------ | ------------ | ------------- | -------------
A | true | 1 | T2 task insertion "happens before" ELT checks the queue, hasTask() == true
B | true | 0 | ELT *may* have processed T2's task it was visible in time, but if not T2 wakes up ELT via eventFdWrite and there will be a "happens before" on `wakenUp` when ELT does a `GAS`
C | false | 1 | Some other thread (e.g. `T1`) already CAS `wakenUp` to `1`. In this scenario because T2 failed to CAS and ELT observed the `wakenUp` there is a "happens before" ELT checks the queue, hasTask() == true
D | false | 0 | ELT did its GAS "early", but will be woken up by some other thread's (e.g. `T1`) call to `Native.eventFdWrite`. When ELT wakes up it will again attempt the GAS and observe the observed the `wakenUp` there is a "happens before" ELT checks the queue, hasTask() == true

If we remove the `CAS` in some scenarios we are dropping some full barriers and ordering constraints provided by JMM on `wakenUp`. If there are alternative means to make the tasks visible (e.g. the queue implementation provides more guarantees than I am aware, or other happens before relationships) that is fine. Just want to make sure things are covered because PR https://github.com/netty/netty/pull/7834 will take away the "max wait" safety net if we miss tasks. Please let me know if I'm missing something.

Reply:
> I don't think JCTools MPSC queue guarantees that insertion from one thread will be visible to another thread immediately

Depends what do you mean by "immediately": i can tell you that isEmpty and poll are consistent with the other concurrent Queues ie isEmpty will return null if there are no pending offers.
So, assuming this as correct (we have many tests on JCtools to stress it), the issue is elsewhere.

Reply:
@Scottmitch you may be thinking of the JCTools `relaxed` method variants

Reply:
@Scottmitch you are correct that the visibility guarantees of JCTools can be surprising, and this is the reason I filed the ticket that started the discussion (#9105), however in this case there is no visibility problem.

`isEmpty()` in JCTools relies on the consumer and producer indexes, the latter of which is maintained by an atomic increment.  Even if the queued item is not yet visible, the fact that there is a queued item arriving _is_ immediately visible.  The event loop may busy-spin waiting for the item when it performs `poll()`, and in doing so `offer()` and `poll()` in JCTools provide volatile visibility semantics also, although at a cost I would prefer we did not incur.

As discussed on #9105 this change _would_ prevent the use of `relaxedPoll()` which does not offer volatile visibility semantics but avoids the busy-spin problem.  But in my opinion, it is incorrect to fix a queue's visibility semantics here, since a queue with volatile visibility semantics is best achieved at the point of the queue itself.  Particularly since we only need a `StoreStore` barrier, not a `CMPXCHG` to achieve this.

Reply:
@franz1981 I think you mean `LoadStore` :)

but yes, you are right, though it doesn't change the point in practice.

Reply:
@belliottsmith ops, you're right...:P